corehq/apps/integration/kyc/services.py

```diff
@@ -22,39 +22,46 @@ def verify_users(kyc_users, config):
     device_id = f'{__name__}.verify_users'
     errors_with_count = defaultdict(int)
     for kyc_user in kyc_users:
-        verification_status = KycVerificationStatus.PENDING
-        try:
-            verification_status = verify_user(kyc_user, config)
-            if verification_status == KycVerificationStatus.FAILED:
-                errors_with_count[KycVerificationFailureCause.USER_INFORMATION_MISMATCH.value] += 1
-        # TODO - Decide on how we want to handle these exceptions for the end user
-        except jsonschema.exceptions.ValidationError:
-            errors_with_count[KycVerificationFailureCause.USER_INFORMATION_INCOMPLETE.value] += 1
-        except (requests.exceptions.ConnectionError, requests.exceptions.Timeout):
-            errors_with_count[KycVerificationFailureCause.NETWORK_ERROR.value] += 1
-        except requests.HTTPError:
-            errors_with_count[KycVerificationFailureCause.API_ERROR.value] += 1
-
-        # Store result only when API successfully returns a verification response
-        if verification_status in (KycVerificationStatus.PASSED, KycVerificationStatus.FAILED):
-            kyc_user.update_verification_status(verification_status, device_id=device_id)
+        verification_status, verification_error = verify_user(kyc_user, config)
+
+        if verification_error:
+            errors_with_count[verification_error] += 1
 
+        kyc_user.update_verification_status(
+            verification_status,
+            device_id=device_id,
+            error_message=verification_error,
+        )
         results[kyc_user.user_id] = verification_status
 
     _report_verification_failure_metric(config.domain, errors_with_count)
     return results
 
 
-def _report_verification_failure_metric(domain, errors_with_count):
-    for error, count in errors_with_count.items():
-        metrics_counter(
-            'commcare.integration.kyc.verification_failure.count',
-            count,
-            tags={'domain': domain, 'error': error}
-        )
+def verify_user(kyc_user, config):
+    verification_error = None
 
+    try:
+        verification_status = _verify_user(kyc_user, config)
+        if verification_status == KycVerificationStatus.FAILED:
+            verification_error = KycVerificationFailureCause.USER_INFORMATION_MISMATCH.value
 
-def verify_user(kyc_user, config):
+    except jsonschema.exceptions.ValidationError:
+        verification_error = KycVerificationFailureCause.USER_INFORMATION_INCOMPLETE.value
+        verification_status = KycVerificationStatus.ERROR
+
+    except (requests.exceptions.ConnectionError, requests.exceptions.Timeout):
+        verification_error = KycVerificationFailureCause.NETWORK_ERROR.value
+        verification_status = KycVerificationStatus.ERROR
+
+    except requests.HTTPError:
```

Comment:
nit: do we need a catch all for other errors? or we intentionally leaving this open since any exception is expected from requests only which seems okay as well.

Reply:
You mean adding an `Exception` block? I'm not sure what errors we expect to get apart from the already-covered ones.